[ML] Restore analytics state if available

dimitris-athanasiou opened the pull request:
This commit restores the model state if available in data
frame analytics jobs.

In addition, this changes the start API so that a stopped job
can be restarted. As we now store the progress in the state index
when the task is stopped, we can use it to determine what state
the job was in when it got stopped.

dimitris-athanasiou committed:
[ML] Restore analytics state if available

This commit restores the model state if available in data
frame analytics jobs.

In addition, this changes the start API so that a stopped job
can be restarted. As we now store the progress in the state index
when the task is stopped, we can use it to determine what state
the job was in when it got stopped.

elasticmachine commented:
Pinging @elastic/ml-core

benwtrent commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java:
```diff
@@ -248,49 +256,94 @@ private void getConfigAndValidate(String id, ActionListener<DataFrameAnalyticsCo
+                        break;
+                    case RESUMING_REINDEXING:
+                    case RESUMING_ANALYZING:
+                        toValidateMappingsListener.onResponse(startContext);
```
Do we need to validate the mappings can be merged if we are already analyzing and have already reindexed all the data?

dimitris-athanasiou replied:
If we are in `resuming_reindexing` we have not necessarily reindexed all the data. In general, I'd rather we perform a redundant check here and keep the path the same rather than skipping it and trying to reason about whether there are some cases that the mappings check would be necessary.

benwtrent commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java:
```diff
@@ -331,6 +384,19 @@ public void onTimeout(TimeValue timeout) {
+            this.config = config;
+            this.progressOnStart = progressOnStart;
+            for (PhaseProgress phase : progressOnStart) {
+                LOGGER.info("[{}] Progress [{}] is [{}]", config.getId(), phase.getPhase(), phase.getProgressPercent());
```
seems like a `debug` level to me?

dimitris-athanasiou replied:
Good catch, that was a leftover. We don't really need it at all.

dimitris-athanasiou commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartDataFrameAnalyticsAction.java:
```diff
@@ -150,12 +153,15 @@ public String toString() {
         static {
             PARSER.declareString(ConstructingObjectParser.constructorArg(), DataFrameAnalyticsConfig.ID);
             PARSER.declareString(ConstructingObjectParser.constructorArg(), DataFrameAnalyticsConfig.VERSION);
+            PARSER.declareObjectArray(ConstructingObjectParser.constructorArg(), PhaseProgress.PARSER, PROGRESS_ON_START);
```
Good catch. This is optional.

dimitris-athanasiou commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartDataFrameAnalyticsAction.java:
```diff
@@ -164,25 +170,36 @@ public static TaskParams fromXContent(XContentParser parser) {
+        public TaskParams(String id, Version version, List<PhaseProgress> progressOnStart) {
             this.id = Objects.requireNonNull(id);
             this.version = Objects.requireNonNull(version);
+            this.progressOnStart = progressOnStart;
```
I've changed the constructor used by the parser to pass an empty list when `progress_on_start` is missing. The main constructor is now calling `Collections.unmodifiableList` which throws if the list is `null`.

dimitris-athanasiou commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartDataFrameAnalyticsAction.java:
```diff
@@ -164,25 +170,36 @@ public static TaskParams fromXContent(XContentParser parser) {
             this.id = in.readString();
             this.version = Version.readVersion(in);
+            if (in.getVersion().onOrAfter(Version.V_7_5_0)) {
+                progressOnStart = in.readList(PhaseProgress::new);
```
On version 7.5.0 onwards it will never be null. I'm addressing the null issues in total.

dimitris-athanasiou commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartDataFrameAnalyticsAction.java:
```diff
@@ -164,25 +170,36 @@ public static TaskParams fromXContent(XContentParser parser) {
         public TaskParams(StreamInput in) throws IOException {
             this.id = in.readString();
             this.version = Version.readVersion(in);
+            if (in.getVersion().onOrAfter(Version.V_7_5_0)) {
```
Not yet. I'll add them right after merging this one in to save the trouble with flipping the versions here.

dimitris-athanasiou committed:
Address review comments

dimitris-athanasiou committed:
Populate progress in TaskParamsTests

dimitris-athanasiou commented:
run elasticsearch-ci/2

dimitris-athanasiou commented:
run elasticsearch-ci/bwc

dimitris-athanasiou commented:
@elasticmachine update branch

dimitris-athanasiou committed:
Improve logging

dimitris-athanasiou committed:
Ensure reindexing progress is at least 1

In order to be able to distinguish between a job
that runs for the first time and another that is restarting,
we ensure reindexing progress is reported to be at least 1
for a running task.

Also fixes an issue in the tests.
